ckan/model/__init__.py

```diff
@@ -266,12 +267,29 @@ def upgrade_db(self, version=None):
 
         @param version: version to upgrade to (if None upgrade to latest)
         '''
+        # pre-upgrade checks
         assert meta.engine.name in ('postgres', 'postgresql'), \
             'Database migration - only Postgresql engine supported (not %s).' \
                 % meta.engine.name
         import migrate.versioning.api as mig
         self.setup_migration_version_control()
         version_before = mig.db_version(self.metadata.bind, self.migrate_repository)
+        from ckan.migration.migrate_package_activity import num_unmigrated
+        # if still at version 0 there can't be any activities needing migrating
+        if version_before > 0:
+            num_unmigrated_dataset_activities = num_unmigrated(meta.engine)
+            if num_unmigrated_dataset_activities:
+                print('''
+    !!! ERROR !!!
+    You have {num_unmigrated} unmigrated package activities.
+
+    You cannot do this db upgrade until you completed the package activity
+    migration first. Full instructions for this situation are here:
+
+    https://github.com/ckan/ckan/wiki/Migrate-package-activity#if-you-tried-to-upgrade-from-ckan-28-or-earlier-to-ckan-29-and-it-stopped-at-paster-db-upgrade
```

Comment:
I thought we could upgrade the db and then migrate the activities later, why is the upgrade being prevented here? Also the `num_unmigrated` implementation walks the whole activity table doing string matching on every record every, that's going to be very slow for large sites.

Reply:
We want people to migrate before they upgrade for the following case: this migration is CKAN 2.9, and then 2.10 comes out which removes revisioning and now migrate_package_activity.py isn't compatible. If a site is upgrading from 2.8 to 2.10 it seems easier if we force people to migrate before upgrade, rather than saying they have to upgrade to 2.9 first, start CKAN then run the migration script and then upgrade to 2.10.

Yeah good point about num_unmigrated. I'll have a ponder about this.

Reply:
To avoid num_unmigrated being run on every upgrade in the future, I've added in migration 88. So when you upgrade to this version of CKAN, it does the check each in the two places, and never again. I seems better than trying to optimize the query or only checking the first package activity. https://github.com/ckan/ckan/pull/3972/commits/b6e72784dc17b42e0bbcffb27bdc901945bca1f5